tcf_website/views/browse.py

```diff
@@ -148,14 +148,70 @@ def course_instructor(request, course_id, instructor_id):
 def instructor_view(request, instructor_id):
     """View for instructor page, showing all their courses taught."""
     instructor = Instructor.objects.get(pk=instructor_id)
-    avg_rating = instructor.average_rating()
-    if avg_rating is not None:
-        avg_rating = round(avg_rating, 2)
-    avg_difficulty = instructor.average_difficulty()
-    if avg_difficulty is not None:
-        avg_difficulty = round(avg_difficulty, 2)
+
+    avg_rating = safe_round(instructor.average_rating())
+    avg_difficulty = safe_round(instructor.average_difficulty())
+
+    courses = group_by_dept(instructor, instructor.get_courses())
     return render(
         request, 'instructor/instructor.html', {
             'instructor': instructor,
             'avg_rating': avg_rating,
-            'avg_difficulty': avg_difficulty})
+            'avg_difficulty': avg_difficulty,
+            'courses': courses
+        })
+
+
+def group_by_dept(instructor, courses):
+    """ Group instructor's courses by subdepartment.
+
+        Returns a dictionary mapping subdepartment names to ids and
+        lists of Course data.
+    """
+    subdept_ids = list(
+        set(courses.values_list('subdepartment', flat=True)))
+
+    # Contains IDs and names of all subdepartments for Instructor's Courses
+    subdepts = sorted(
+        list(
+            map(
+                lambda subdept_id: [
+                    subdept_id,
+                    Subdepartment.objects.get(
+                        pk=subdept_id).name],
+                subdept_ids)),
+        key=lambda x: x[1])
+
+    # Create dictionary corresponding dept to courses
+    grouped_courses = {}
+    for subdept in subdepts:
+        grouped_courses[subdept[1]] = {
+            "id": subdept[0],
+            "courses": []
+        }
+
+    for course in courses:
+        course_subdept = course.subdepartment.name
+        course_data = {
+            'name': str(course), 'id': course.id, 'avg_rating': safe_round(
+                Instructor.average_rating_for_course(
+                    instructor, course)), 'avg_difficulty': safe_round(
+                Instructor.average_difficulty_for_course(
+                    instructor, course)), 'avg_hours': (
+                        lambda x: safe_round(x) if x is not None else '\u2014')(
```

Comment:
Could we just delete this part since `safe_round` already does error checking
```suggestion
                        lambda x: safe_round(x)) (
```

Reply:
Originally there was separate code because I wanted to truncate hours and not round to 2 decimal points, but then the decision to keep precision on hours as well rendered this invalid. Good catch, but I guess since we're removing hours entirely it doesn't matter anymore.